### PR TITLE
Add additional request tracking

### PR DESCRIPTION
**Pull request by AlanFoster**
*Note* The first commit is not related to this PR. Once https://github.com/oesmith/puffing-billy/pull/224 is in, the first commit will disappear.

Closes https://github.com/oesmith/puffing-billy/issues/195

### What's changed?

This pull request allow users to track all requests that have gone through Puffing Billy. I decided to leave Puffing Billy un-opinionated here, and just track requests. i.e. It gives all data to the user, and they can decide if they wish to fail a build if there's unhandled web requests etc.

An example configuration to use this may be:

```ruby
Billy.configure do |c|
  c.cache = false
  c.record_stub_requests = true
  c.record_requests = true
  c.non_whitelisted_requests_disabled = true
end
```

### Application

With Puffing Billy now exposing the tracked requests via its API, I'm now able to do some really powerful things. For instance, I now have two post RSpec hooks:

- Failing a test if there are any unhandled requests. I now fail the RSpec test if there's any unhandled web requests via puffing billy.
- Logging Additional information during an exception. We have intermittents on our CI environments, and I would like to know what network requests were made via puffing billy. This can be a useful source of information as it can help pinpoint a race condition. In this scenario I have written a custom Exception that wraps the original RSepec `example.exception`, and additionally spits out all requests as a table. This can be seen below.

image

In the below scenario I can immediately work out that I haven't handled the OPTIONs request correctly, and that has subsequently led to a Capybara-Webkit crash. With the printed table, containing the URL, the fix is now immediately obvious:

```diff
- proxy.stub('https://example.com/api', :method => :options).and_return(
+ proxy.stub('https://example.com:443/api', :method => :options).and_return(
```

### CODE REVIEW

File: lib/billy/handlers/request_handler.rb [hunk 1]
```diff
@@ -14,31 +15,42 @@ def handlers
...
       # Process the handlers by order of importance
       [:stubs, :cache, :proxy].each do |key|
         if (response = handlers[key].handle_request(method, url, headers, body))
+          @request_log.complete(request, key)
           return response
         end
       end
 
       body_msg = Billy.config.cache_request_body_methods.include?(method) ? " with body '#{body}'" : ''
+      request_log.complete(request, :error)
```
Comment (AlanFoster) [inline lib/billy/handlers/request_handler.rb]: I'm currently using this information to spit out a nice table via TablePrint:

```ruby
    TablePrint::Printer.table_print(requests, [
      :status,
      :handler,
      :method,
      { url: { width: 150 } },
      :headers,
      :body
    ])
```

This gives me a nice table such as:

```
STATUS   | HANDLER | METHOD  | URL                            | HEADERS                        | BODY
---------|---------|---------|--------------------------------|--------------------------------|-----
complete | proxy   | GET     | http://127.0.0.1:9887/example  | {"Accept"=>"text/html,appli... |
complete | proxy   | GET     | http://127.0.0.1:9887/example  | {"Accept"=>"text/css,*/*;q=... |
complete | proxy   | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
complete | proxy   | GET     | http://127.0.0.1:9887/example  | {"User-Agent"=>"Mozilla/5.0... |
complete | proxy   | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
inflight |         | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
inflight |         | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
inflight |         | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
inflight |         | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
inflight |         | GET     | http://127.0.0.1:9887/example  | {"Referer"=>"http://127.0.0... |
complete | error   | OPTIONS | https://example.com:443/exa... | {"Access-Control-Request-Me... |
```

I can immediately read `:error` from this output, but perhaps it's not semantically accurate. Should it maybe be `:unhandled` instead ?

Any preferences here @ronwsmith 🤔

Comment (ronwsmith) [reply lib/billy/handlers/request_handler.rb]: I'm okay with `:error` since we're using `error` for the below logging line already.

**Comment by AlanFoster** [comment]
@ronwsmith Updated! :)

I've rebased this, resolved conflicts, and now the readme contains examples of the `record_requests` configuration, as well as a potential usage of logging all requests.

I have a slightly different setup for my own tests, where it only outputs the puffing billy request logs on test failure - but I think that's okay to skip out in this documentation 🤔

**Comment by AlanFoster** [comment]
@ronwsmith Let me know if there's anything more I can do to get this delivered into master :+1:

**Comment by ronwsmith** [comment]
Looks good, I'll try to get this in next week.

**Comment by ronwsmith** [comment]
Released in 1.1.0

**Pull request merged**